File: src/dialogs/settingswindow.cpp
#include <QColorDialog>
#include <QFileDialog>
#include <QLineEdit>
#include <QListWidget>
#include <QListWidgetItem>
#include <QPixmap>
#include <QPushButton>
#include <QSettings>
#include <QStyle>

#include "mainwindow.h"
#include "dialogs/settingswindow.h"
#include "ui_settingswindow.h"
#include "util.h"

SettingsWindow::SettingsWindow(QWidget *parent) : QDialog(parent), ui(new Ui::SettingsWindow) {
	ui->setupUi(this);
	ui->neosphereDir_btn->setIcon(style()->standardIcon(QStyle::SP_DialogOpenButton));
	ui->legacySphereDir_btn->setIcon(style()->standardIcon(QStyle::SP_DialogOpenButton));
	m_currentTheme = MainWindow::instance()->getTheme();
	m_newTheme = m_currentTheme;
	QSettings settings;
	int size = settings.beginReadArray("projectDirs");
	for(int i = 0; i < size; ++i) {
		settings.setArrayIndex(i);
		QListWidgetItem* item = new QListWidgetItem(settings.value("directory").toString(), ui->projectDirsList);
		item->setFlags(item->flags()|Qt::ItemIsEditable);
		ui->projectDirsList->addItem(item);
	}
	settings.endArray();
	if(settings.value("whichEngine", "neosphere") == "legacy") {
		ui->whichEngine_combo->setCurrentIndex(1);
	} else {
		ui->whichEngine_combo->setCurrentIndex(0);
	}
	ui->neosphereDir_txt->setText(settings.value("neosphereDir").toString());
	ui->legacySphereDir_txt->setText(settings.value("legacySphereDir").toString());
	m_mapCursorColor = settings.value("mapCursorColor", "#0080ff").toString();
	m_gridColor = settings.value("gridColor", "#000000").toString();
	QColor col(m_mapCursorColor);
	if(col.isValid()) ui->mapCursorCol_btn->setColor(col);
	col = QColor::fromString(m_gridColor);
	if(col.isValid()) ui->gridColor_btn->setColor(col);
	connect(ui->mapCursorCol_btn,SIGNAL(colorChanged(QColor)),this,SLOT(mapCursorColChanged(QColor)));
	connect(ui->gridColor_btn, SIGNAL(colorChanged(QColor)), this, SLOT(gridColorChanged(QColor)));
	connect(ui->buttonBox->button(QDialogButtonBox::Ok), SIGNAL(clicked()), this, SLOT(onOk()));
	connect(ui->buttonBox->button(QDialogButtonBox::Apply), SIGNAL(clicked()), this, SLOT(onApply()));
	connect(ui->buttonBox->button(QDialogButtonBox::Cancel), SIGNAL(clicked()), this, SLOT(onCancel()));
	connect(ui->neosphereDir_btn, &QToolButton::clicked, this, &SettingsWindow::onNeosphereDirButtonClicked);
	connect(ui->legacySphereDir_btn, &QToolButton::clicked, this, &SettingsWindow::onLegacySphereDirButtonClicked);
	connect(ui->themeCombo, &QComboBox::currentTextChanged, this, &SettingsWindow::onThemeComboCurrentTextChanged);
	connect(ui->addDirButton, &QToolButton::clicked, this, &SettingsWindow::onAddDirButtonClicked);
	connect(ui->removeDirButton, &QToolButton::clicked, this, &SettingsWindow::onRemoveDirButtonClicked);
	connect(ui->browseDirButton, &QToolButton::clicked, this, &SettingsWindow::onBrowseDirButtonClicked);
}

SettingsWindow::~SettingsWindow() {
	disconnect(ui->mapCursorCol_btn,SIGNAL(colorChanged(QColor)),this,SLOT(mapCursorColChanged(QColor)));
	disconnect(ui->gridColor_btn, SIGNAL(colorChanged(QColor)), this, SLOT(gridColorChanged(QColor)));
	disconnect(ui->buttonBox->button(QDialogButtonBox::Ok), SIGNAL(clicked()), this, SLOT(onOk()));
	disconnect(ui->buttonBox->button(QDialogButtonBox::Apply), SIGNAL(clicked()), this, SLOT(onApply()));
	disconnect(ui->buttonBox->button(QDialogButtonBox::Cancel), SIGNAL(clicked()), this, SLOT(onCancel()));
	disconnect(ui->neosphereDir_btn, &QToolButton::clicked, this, &SettingsWindow::onNeosphereDirButtonClicked);
	disconnect(ui->legacySphereDir_btn, &QToolButton::clicked, this, &SettingsWindow::onLegacySphereDirButtonClicked);
	disconnect(ui->themeCombo, &QComboBox::currentTextChanged, this, &SettingsWindow::onThemeComboCurrentTextChanged);
	disconnect(ui->addDirButton, &QToolButton::clicked, this, &SettingsWindow::onAddDirButtonClicked);
	disconnect(ui->removeDirButton, &QToolButton::clicked, this, &SettingsWindow::onRemoveDirButtonClicked);
	disconnect(ui->browseDirButton, &QToolButton::clicked, this, &SettingsWindow::onBrowseDirButtonClicked);
	delete ui;
}

void SettingsWindow::onOk() {
	m_currentTheme = m_newTheme;
	saveSettings();
	accept();
}

void SettingsWindow::onApply() {
	m_currentTheme = m_newTheme;
	saveSettings();
}

void SettingsWindow::onCancel() {
	MainWindow::instance()->setTheme(m_currentTheme);
	reject();
}

void SettingsWindow::onNeosphereDirButtonClicked() {
	QString _m_neosphereDir = QFileDialog::getExistingDirectory(this, "Open Directory");
	if(_m_neosphereDir == "") return;
	m_neosphereDir = _m_neosphereDir;
	ui->neosphereDir_txt->setText(m_neosphereDir);
}

void SettingsWindow::onLegacySphereDirButtonClicked() {
	QString _m_legacySphereDir = QFileDialog::getExistingDirectory(this, "Open Directory", m_legacySphereDir);
	if(_m_legacySphereDir == "") return;
	m_legacySphereDir = _m_legacySphereDir;
	ui->legacySphereDir_txt->setText(m_legacySphereDir);
}

void SettingsWindow::onThemeComboCurrentTextChanged(const QString &newtext) {
	if(newtext == "Default")
		m_newTheme = "";
	else if(newtext == "Dark")
		m_newTheme = ":/text/dark.qss";
	else
		m_newTheme = "";
	MainWindow::instance()->setTheme(m_newTheme);
}

void SettingsWindow::saveSettings() {
	QSettings settings;
	int numDirs = ui->projectDirsList->count();
	settings.remove("projectDirs");
	settings.beginWriteArray("projectDirs");
	for(int d = 0; d < numDirs; d++) {
		settings.setArrayIndex(d);
		QString directory = ui->projectDirsList->item(d)->text();
		if(directory != "" && !ui->projectDirsList->item(d)->isHidden()) {
			settings.setValue("directory", directory);
		}
	}
	settings.endArray();
	settings.setValue("theme", m_currentTheme);
	if(ui->whichEngine_combo->currentText() == "Use Sphere 1.x (legacy)") {
		settings.setValue("whichEngine", "legacy");
	} else {
		settings.setValue("whichEngine", "neosphere");
	}
	MainWindow::instance()->setEngine(settings.value("whichEngine").toString());
	settings.setValue("neosphereDir", ui->neosphereDir_txt->text());
	settings.setValue("legacySphereDir", ui->legacySphereDir_txt->text());
	settings.setValue("mapCursorColor", m_mapCursorColor);
	settings.setValue("gridColor", m_gridColor);
}

void SettingsWindow::onAddDirButtonClicked() {
	QListWidgetItem* newItem = new QListWidgetItem("", ui->projectDirsList);
	newItem->setFlags(newItem->flags()|Qt::ItemIsEditable);
	ui->projectDirsList->addItem(newItem);
	ui->projectDirsList->setCurrentRow(ui->projectDirsList->count()-1);
}

void SettingsWindow::onRemoveDirButtonClicked() {
	ui->projectDirsList->takeItem(ui->projectDirsList->currentRow());
}

void SettingsWindow::mapCursorColChanged(QColor color) {
	m_mapCursorColor = color.name();
}

void SettingsWindow::gridColorChanged(QColor color) {
	m_gridColor = color.name();
}

void SettingsWindow::onBrowseDirButtonClicked() {
	int row = ui->projectDirsList->currentRow();
	if(row < 0) return;
	QString dir = QFileDialog::getExistingDirectory(this, "Open Directory");
	if(dir != "") {
		ui->projectDirsList->item(row)->setText(dir);
	}
}

File: dialogs/settingswindow.py
from enum import Enum
import os

from PySide6.QtCore import Qt, Slot, Signal
from PySide6.QtGui import QColor, QGuiApplication
from PySide6.QtWidgets import QDialog, QWidget, QDialogButtonBox, QListWidgetItem, QFileDialog, QMenu, QMessageBox

from settings import Settings, Defaults

from widgets.colorbutton import ColorButton

from ui.ui_settingswindow import Ui_SettingsWindow

class UnrecognizedFileHandler(Enum):
	ExternalEditor = 0,
	InternalTextEditor = 1

class SettingsWindow(QDialog):
	ui: Ui_SettingsWindow
	settingsSaved: Signal = Signal()
	def __init__(self, parent: QWidget|None = None, wtype: Qt.WindowType = Qt.WindowType.Widget):
		super().__init__(parent, wtype)
		self.ui = Ui_SettingsWindow()
		self.ui.setupUi(self)

		self.ui.buttonBox.button(QDialogButtonBox.StandardButton.Ok).clicked.connect(self.onOK)
		self.ui.buttonBox.button(QDialogButtonBox.StandardButton.Apply).clicked.connect(self.onApply)
		self.ui.buttonBox.button(QDialogButtonBox.StandardButton.Cancel).clicked.connect(self.reject)
		self.ui.addDirButton.clicked.connect(self.onAddDirButtonClicked)
		self.ui.removeDirButton.clicked.connect(self.onRemoveDirButtonClicked)
		self.ui.browseDirButton.clicked.connect(self.onBrowseDirButtonClicked)
		self.ui.neosphereDir_btn.clicked.connect(self.onNeoSphereDirButtonClicked)
		self.ui.legacySphereDir_btn.clicked.connect(self.onLegacySphereDirButtonClicked)
		
		self._addColorMenu(self.ui.gridColor_btn)
		self._addColorMenu(self.ui.mapCursorCol_btn)
		self._removeWineIfWindows()
		self.loadSettings()

	def _addColorMenu(self, btn:ColorButton):
		menu = QMenu(btn)
		menu.addAction("Copy color to clipboard").triggered.connect(lambda: QGuiApplication.clipboard().setText(btn.color.name()))
		menu.addAction("Reset color").triggered.connect(lambda: self._resetColorPressed(btn))
		btn.customContextMenuRequested.connect(lambda pt: menu.exec(btn.mapToGlobal(pt)))
	
	def loadSettings(self):
		settings = Settings()
		self.ui.mapCursorCol_btn.setColor(QColor(settings.mapCursorColor))
		self.ui.gridColor_btn.setColor(QColor(settings.gridColor))
		if os.name != "nt":
			self.ui.wineDir_txt.setText(settings.wineDir or "/usr/bin")
		self.ui.neosphereDir_txt.setText(settings.neosphereDir or "")
		self.ui.legacySphereDir_txt.setText(settings.legacySphereDir or "")
		self.ui.unrecognizedFileEditor_combo.setCurrentIndex(0 if settings.unrecognizedFileEditor == "external" else 1)
		self.ui.whichEngine_combo.setCurrentIndex(1 if settings.whichEngine == "legacy" else 0)

		self.ui.projectDirsList.clear()
		searchPaths = settings.projectDirs
		for sPath in searchPaths:
			if sPath != "":
				self._addProjectDirItem(sPath)

	def _saveSettings(self):
		settings = Settings()
		settings.mapCursorColor = self.ui.mapCursorCol_btn.color
		settings.gridColor = self.ui.gridColor_btn.color
		if os.name != "nt":
			settings.wineDir = self.ui.wineDir_txt.text()
		settings.neosphereDir = self.ui.neosphereDir_txt.text()
		settings.legacySphereDir = self.ui.legacySphereDir_txt.text()
		settings.unrecognizedFileEditor = "text" if self.ui.unrecognizedFileEditor_combo.currentIndex() == 1 else "external"
		settings.whichEngine = "legacy" if self.ui.whichEngine_combo.currentIndex() == 1 else "neosphere"

		del settings.projectDirs
		numDirs = self.ui.projectDirsList.count()
		dirs = []
		for d in range(numDirs):
			directory = self.ui.projectDirsList.item(d).text()
			if directory != "" and not self.ui.projectDirsList.item(d).isHidden():
				dirs.append(directory)
		settings.projectDirs = dirs
		self.settingsSaved.emit()

	def _addProjectDirItem(self, text:str):
		item = QListWidgetItem(text, self.ui.projectDirsList)
		item.setFlags(item.flags()|Qt.ItemFlag.ItemIsEditable)
		self.ui.projectDirsList.addItem(item)

	def _removeWineIfWindows(self):
		if os.name == "nt":
			# running in Windows, no need for WINE stuff
			settings = Settings()
			del settings.wineDir
			self.layout().removeWidget(self.ui.wineDir_lbl)
			self.ui.wineDir_layout.removeWidget(self.ui.wineDir_btn)
			self.ui.wineDir_layout.removeWidget(self.ui.wineDir_txt)
			self.ui.wineDir_lbl.deleteLater()
			self.ui.wineDir_btn.deleteLater()
			self.ui.wineDir_txt.deleteLater()
			self.ui.wineDir_layout.deleteLater()

	def _resetColorPressed(self, btn:ColorButton):
		match btn:
			case self.ui.mapCursorCol_btn:
				btn.setColor(Defaults.mapCursorColor.value)
			case self.ui.gridColor_btn:
				btn.setColor(Defaults.gridColor.value)
			case _:
				QMessageBox.critical(self, "Error", f"Unrecognized color button '{btn.objectName()}'")

	@Slot()
	def onOK(self):
		self._saveSettings()
		self.accept()

	@Slot()	
	def onApply(self):
		self._saveSettings()

	@Slot()
	def onAddDirButtonClicked(self):
		self._addProjectDirItem("")
		self.ui.projectDirsList.setCurrentRow(self.ui.projectDirsList.count()-1)

	@Slot()
	def onRemoveDirButtonClicked(self):
		self.ui.projectDirsList.takeItem(self.ui.projectDirsList.currentRow())

	@Slot()
	def onBrowseDirButtonClicked(self):
		row = self.ui.projectDirsList.currentRow()
		if row < 0:
			return
		dir = QFileDialog.getExistingDirectory(self, "Open Directory")
		if dir != "":
			self.ui.projectDirsList.item(row).setText(dir)

	@Slot()
	def onNeoSphereDirButtonClicked(self):
		dir = QFileDialog.getExistingDirectory(self, "Open Directory")
		if dir != "":
			self.ui.neosphereDir_txt.setText(dir)

	@Slot()
	def onLegacySphereDirButtonClicked(self):
		dir = QFileDialog.getExistingDirectory(self, "Open Directory")
		if dir != "":
			self.ui.legacySphereDir_txt.setText(dir)
